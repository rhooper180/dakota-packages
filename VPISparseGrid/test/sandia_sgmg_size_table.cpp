# include "sandia_rules.hpp"
# include "sandia_rules2.hpp"
# include "sandia_sgmg.hpp"

# include <cstdlib>
# include <iostream>
# include <iomanip>
# include <cmath>
//
//  Two global variables needed to support the "parameter" function.
//
double *P;
int *NP;

int main ( );
void sgmg_size_tabulate 
( 
  int gw_compute_order_1d ( int level, int growth ), 
  int growth_1d,
  int np_1d,
  double p_1d[], 
  void gw_compute_points_1d ( int order, int dim, double w[] ),
  int dim_min,
  int dim_max,
  int level_max_min,
  int level_max_max 
);

typedef void ( *GWPointer ) ( int order, int dim, double w[] );
typedef int ( *GWPointer2 ) ( int level, int growth );

//****************************************************************************80

int main ( )

//****************************************************************************80
//
//  Purpose:
//
//    MAIN makes a point count table.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license.
//
//  Modified:
//
//    01 January 2012
//
//  Author:
//
//    John Burkardt
//
//  Reference:
//
//    Fabio Nobile, Raul Tempone, Clayton Webster,
//    A Sparse Grid Stochastic Collocation Method for Partial Differential
//    Equations with Random Input Data,
//    SIAM Journal on Numerical Analysis,
//    Volume 46, Number 5, 2008, pages 2309-2345.
//
{
  double ctime;
  int dim_max;
  int dim_min;
  int growth_1d;
  int level_max_max;
  int level_max_min;
  int np_1d;
  double *p_1d;

  webbur::timestamp ( );
  std::cout << "\n";
  std::cout << "SGMG_SIZE_TABLE\n";
  std::cout << "  C++ version\n";
//
//  Clenshaw-Curtis, Slow Exponential Growth.
//
  if ( true )
  {
    growth_1d = 0;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 1;
    dim_max = 5;
    level_max_min = 0;
    level_max_max = 7;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_cc, growth_1d, np_1d, 
      p_1d, webbur::clenshaw_curtis_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Clenshaw-Curtis, Slow Exponential Growth.
//
  if ( true )
  {
    growth_1d = 0;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 6;
    dim_max = 10;
    level_max_min = 0;
    level_max_max = 5;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_cc, growth_1d, np_1d, 
      p_1d, webbur::clenshaw_curtis_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Clenshaw-Curtis Grid, Full exponential growth.
//
  if ( true )
  {
    growth_1d = 2;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 1;
    dim_max = 5;
    level_max_min = 0;
    level_max_max = 7;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_cc, growth_1d, np_1d, 
      p_1d, webbur::clenshaw_curtis_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Clenshaw-Curtis Grid, Full exponential growth.
//
  if ( true )
  {
    growth_1d = 2;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 6;
    dim_max = 10;
    level_max_min = 0;
    level_max_max = 5;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_cc, growth_1d, np_1d, 
      p_1d, webbur::clenshaw_curtis_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Clenshaw-Curtis Grid, Full exponential growth.
//
  if ( true )
  {
    growth_1d = 2;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 100;
    dim_max = 100;
    level_max_min = 0;
    level_max_max = 2;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_cc, growth_1d, np_1d, 
      p_1d, webbur::clenshaw_curtis_points, 
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Patterson, Slow Exponential Growth.
//
  if ( true )
  {
    growth_1d = 0;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 1;
    dim_max = 5;
    level_max_min = 0;
    level_max_max = 7;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_gp, growth_1d, np_1d, 
      p_1d, webbur::patterson_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Patterson, Slow Exponential Growth.
//
  if ( true )
  {;
    growth_1d = 0;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 6;
    dim_max = 10;
    level_max_min = 0;
    level_max_max = 5;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_gp, growth_1d, np_1d, 
      p_1d, webbur::patterson_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Patterson, Moderate Exponential Growth.
//
  if ( true )
  {
    growth_1d = 1;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 1;
    dim_max = 5;
    level_max_min = 0;
    level_max_max = 7;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_gp, growth_1d, np_1d, 
      p_1d, webbur::patterson_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Patterson, Moderate Exponential Growth.
//
  if ( true )
  {
    growth_1d = 1;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 6;
    dim_max = 10;
    level_max_min = 0;
    level_max_max = 5;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_gp, growth_1d, np_1d, 
      p_1d, webbur::patterson_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Patterson, Full Exponential Growth
//
  if ( true )
  {
    growth_1d = 2;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 1;
    dim_max = 5;
    level_max_min = 0;
    level_max_max = 7;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_gp, growth_1d, np_1d, 
      p_1d, webbur::patterson_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Patterson, Full Exponential Growth.
//
  if ( true )
  {
    growth_1d = 2;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 6;
    dim_max = 10;
    level_max_min = 0;
    level_max_max = 5;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_exp_gp, growth_1d, np_1d, 
      p_1d, webbur::patterson_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Gauss-Legendre, slow linear odd growth.
//
  if ( true )
  {
    growth_1d = 0;
    np_1d = 0;
    p_1d = new double[np_1d];
    dim_min = 1;
    dim_max = 5;
    level_max_min = 0;
    level_max_max = 7;
    ctime = webbur::cpu_time ( );
    sgmg_size_tabulate ( webbur::level_to_order_linear_wn, growth_1d, np_1d, 
      p_1d, webbur::legendre_points,
      dim_min, dim_max, level_max_min, level_max_max );
    ctime = webbur::cpu_time ( ) - ctime;
    std::cout << "\n";
    std::cout << "  CPU Time = " << ctime << "\n";
    delete [] p_1d;
  }
//
//  Terminate.
//
  std::cout << "\n";
  std::cout << "SGMG_SIZE_TABLE\n";
  std::cout << "  Normal end of execution.\n";

  std::cout << "\n";
  webbur::timestamp ( );
  
  return 0;
}
namespace webbur 
{
//****************************************************************************80

double parameter ( int dim, int offset )

//****************************************************************************80
//
//  Purpose:
//
//    PARAMETER is a user-supplied routine to retrieve parameters.
//
//  Discussion:
//
//    The declaration for this function is in SANDIA_RULES.H
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license. 
//
//  Modified:
//
//    02 April 2010
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, int DIM, the spatial dimension.
//
//    Input, int OFFSET, the offset of the parameter within the 
//    spatial dimension.
//
//    Output, double PARAMETER, the value of the OFFSET-th parameter
//    associated with the DIM-th dimension.
//
{
  int i;
  int j;
  double value;

  j = 0;
  for ( i = 0; i < dim; i++ )
  {
    j = j + NP[i];
  }
  value = P[j+offset];

  return value;
}
}
//***************************************************************************80

void sgmg_size_tabulate 
( 
  int gw_compute_order_1d ( int level, int growth ), 
  int growth_1d,
  int np_1d,
  double p_1d[], 
  void gw_compute_points_1d ( int order, int dim, double w[] ),
  int dim_min,
  int dim_max,
  int level_max_min,
  int level_max_max 
)

//***************************************************************************80
//
//  Purpose:
//
//    SGMG_SIZE_TABULATE tests SGMG_SIZE.
//
//  Discussion:
//
//    We do NOT consider mixed rules.  Instead, we are looking at sparse grid
//    rules for which all dimensions use the same 1D rule family.
//
//  Licensing:
//
//    This code is distributed under the GNU LGPL license. 
//
//  Modified:
//
//    01 January 2012
//
//  Author:
//
//    John Burkardt
//
//  Parameters:
//
//    Input, int GW_COMPUTE_ORDER_1D ( int level, int growth ),
//    a function which returns the order of the
//    1D quadrature rule of a given level and growth rule.
//
//    Input, int GROWTH, the growth rule. 
//    0, slow;
//    1, moderate;
//    2, full.
//
//    Input, int NP_1D, the number of parameters in the 1D rule.
//
//    Input, double P_1D[NP_1D], the parameters for the 1D rule.
//
//    Input, void GW_COMPUTE_POINTS_1D ( int order, int dim, double w[] ),
//    the function to be used to compute points for the 1D rule.
//
//    Input, int DIM_MIN, the minimum spatial dimension.
//
//    Input, int DIM_MAX, the maximum spatial dimension.
//
//    Input, int LEVEL_MAX_MIN, the minimum value of LEVEL_MAX.
//
//    Input, int LEVEL_MAX_MAX, the maximum value of LEVEL_MAX.
//
{
  int dim;
  int dim_num;
  int growth;
  GWPointer2 *gw_compute_order;
  GWPointer *gw_compute_points;
  int i;
  int j;
  int level_max;
  int *np;
  int np_sum;
  double *p;
  int point_num;
  double tol;

  std::cout << "\n";
  std::cout << "SGMG_SIZE_TABULATE\n";
  std::cout << "  SGMG_SIZE returns the number of distinct\n";
  std::cout << "  points in a sparse grid.\n";
  std::cout << "\n";
  std::cout << "  We use the same rule in all dimensions, and count the points\n";
  std::cout << "  for a range of dimensions and levels.\n";
  std::cout << "\n";

  tol = std::sqrt ( webbur::r8_epsilon ( ) );

  std::cout << "   DIM: ";
  for ( dim_num = dim_min; dim_num <= dim_max; dim_num++ )
  {
    std::cout << "  " << std::setw(8) << dim_num;
  }
  std::cout<< "\n";
  std::cout << "\n";
  std::cout << "   LEVEL_MAX\n";
  std::cout << "\n";

  for ( level_max = level_max_min; level_max <= level_max_max; level_max++ )
  {
    std::cout << "    " << std::setw(4) << level_max;

    for ( dim_num = dim_min; dim_num <= dim_max; dim_num++ )
    {
      growth = growth_1d;
      NP = new int[dim_num];
      np_sum = dim_num * np_1d;
      P = new double[np_sum];
      gw_compute_points = new GWPointer[dim_num];
      gw_compute_order = new GWPointer2[dim_num];

      j = 0;
      for ( dim = 0; dim < dim_num; dim++ )
      {
        NP[dim] = np_1d;
        for ( i = 0; i < np_sum; i++ )
        {
          P[j] = p_1d[i];
        }
        gw_compute_points[dim] = gw_compute_points_1d;
        gw_compute_order[dim] = gw_compute_order_1d;
      }

      point_num = webbur:: sgmg_size ( dim_num, level_max, gw_compute_points, 
        tol, growth, gw_compute_order );

      std::cout << "  " << std::setw(6) << point_num;

      delete [] gw_compute_order;
      delete [] gw_compute_points;
      delete [] NP;
      delete [] P;
    }
    std::cout << "\n";
  }
  return;
}
