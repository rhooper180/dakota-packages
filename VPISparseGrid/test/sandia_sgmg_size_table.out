06 January 2012 10:36:30 AM

SGMG_SIZE_TABLE
  C++ version

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          1         2         3         4         5

   LEVEL_MAX

       0       1       1       1       1       1
       1       3       5       7       9      11
       2       5      13      25      41      61
       3       9      29      69     137     241
       4       9      49     153     369     761
       5      17      81     297     849    2033
       6      17     129     545    1777    4833
       7      17     161     881    3377   10433

  CPU Time = 0.184414

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          6         7         8         9        10

   LEVEL_MAX

       0       1       1       1       1       1
       1      13      15      17      19      21
       2      85     113     145     181     221
       3     389     589     849    1177    1581
       4    1409    2409    3873    5929    8721
       5    4289    8233   14689   24721   39665

  CPU Time = 0.926352

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          1         2         3         4         5

   LEVEL_MAX

       0       1       1       1       1       1
       1       3       5       7       9      11
       2       5      13      25      41      61
       3       9      29      69     137     241
       4      17      65     177     401     801
       5      33     145     441    1105    2433
       6      65     321    1073    2929    6993
       7     129     705    2561    7537   19313

  CPU Time = 0.226

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          6         7         8         9        10

   LEVEL_MAX

       0       1       1       1       1       1
       1      13      15      17      19      21
       2      85     113     145     181     221
       3     389     589     849    1177    1581
       4    1457    2465    3937    6001    8801
       5    4865    9017   15713   26017   41265

  CPU Time = 0.935453

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:        100

   LEVEL_MAX

       0       1
       1     201
       2   20201

  CPU Time = 0.467776

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          1         2         3         4         5

   LEVEL_MAX

       0       1       1       1       1       1
       1       3       5       7       9      11
       2       3       9      19      33      51
       3       7      17      39      81     151
       4       7      33      87     193     391
       5       7      33     135     385     903
       6      15      65     207     641    1743
       7      15      97     399    1217    3343

  CPU Time = 0.092519

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          6         7         8         9        10

   LEVEL_MAX

       0       1       1       1       1       1
       1      13      15      17      19      21
       2      73      99     129     163     201
       3     257     407     609     871    1201
       4     737    1303    2177    3463    5281
       5    1889    3655    6657   11527   19105

  CPU Time = 0.667606

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          1         2         3         4         5

   LEVEL_MAX

       0       1       1       1       1       1
       1       3       5       7       9      11
       2       7      17      31      49      71
       3      15      49     111     209     351
       4      15      97     303     705    1391
       5      15     161     687    1985    4623
       6      31     257    1375    4865   13343
       7      31     321    2335   10369   33823

  CPU Time = 0.40472

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          6         7         8         9        10

   LEVEL_MAX

       0       1       1       1       1       1
       1      13      15      17      19      21
       2      97     127     161     199     241
       3     545     799    1121    1519    2001
       4    2465    4047    6273    9295   13281
       5    9377   17263   29569   47887   74145

  CPU Time = 1.34322

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          1         2         3         4         5

   LEVEL_MAX

       0       1       1       1       1       1
       1       3       5       7       9      11
       2       7      17      31      49      71
       3      15      49     111     209     351
       4      31     129     351     769    1471
       5      63     321    1023    2561    5503
       6     127     769    2815    7937   18943
       7     255    1793    7423   23297   61183

  CPU Time = 0.540241

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          6         7         8         9        10

   LEVEL_MAX

       0       1       1       1       1       1
       1      13      15      17      19      21
       2      97     127     161     199     241
       3     545     799    1121    1519    2001
       4    2561    4159    6401    9439   13441
       5   10625   18943   31745   50623   77505

  CPU Time = 1.35658

SGMG_SIZE_TABULATE
  SGMG_SIZE returns the number of distinct
  points in a sparse grid.

  We use the same rule in all dimensions, and count the points
  for a range of dimensions and levels.

   DIM:          1         2         3         4         5

   LEVEL_MAX

       0       1       1       1       1       1
       1       3       5       7       9      11
       2       3       9      19      33      51
       3       5      17      39      81     151
       4       5      33      87     193     391
       5       7      45     153     409     933
       6       7      81     273     777    1973
       7       9      97     465    1481    4013

  CPU Time = 0.071574

SGMG_SIZE_TABLE
  Normal end of execution.

06 January 2012 10:36:38 AM
